# 1 原因(内因和外因)
A 内因
代码不稳健. 这只能积累测试,提高稳定性(健壮性)
B 外因
一般是意外. 比如文件被删除,做好风险适应性

异常(Exception). 异常也是类, 有BaseException和Exception; 前者是异常基类
后者是常规异常类. TypeError NameError 等等


异常的解决: (1)自己解决: 流程控制(判断语句)
           (2) 别人解决:抛出异常
 


# 2 处理方法1: 针对原因1 (流程上控制_判断语句)
try :
    可能会异常的代码
except(异常) as  e:
    解决异常的代码
else:
   没有异常就执行的代码
 finally:
   不管有没有异常都执行,先于return 执行


# 3 处理方法2: 针对原因2
raise PermissionError
考虑完善代码
# 4 自定义异常
例子: 比如用户名必须长度4-8,且纯数字
" 1 用户输入注册信息
  2 判断是否纯数
  3 判断长度
  4 注册成功
" 
自定义异常类
`class CustomExcetion(Exception):
 *初始化对象属性
 def __init__(self,*args, ** kargs):
     super().__init__(*args,**kags)
 
 * 模拟注册
def register():
   str_name=input()
   * 判断是否为纯数
   flag = str_name.isdigit()
    * 用flag来判断
    if flag:
       * 判断长度4-8, 如果成立,抛出异常
       if 0==len(str_name) or 3>=len(str_name) or 9<len(str_name):
          raise CustomException('用户名不合法,输入长度4-8')
       
    else:
        raise CustomException('用户名不合法,输入纯数字')
    
    * 提示
    print(f'注册成功')`
    
try:
   register()
except  CustomException as e:
   print('程序异常,异常为{e}')
   
[参见](https://www.bilibili.com/video/BV1Wv411s7pF?p=81)
   
     
   
   
